### PR TITLE
Movement of Cows bug

### PR DESCRIPTION
**Pull request by thomrawson**
tests/testthat/movement_matrix_bug.R is a script that will reproduce and demonstrate a bug I've noticed.

The script runs the "actual" model of, herds in 48 US regions.

The trajectory for total susceptible cows in region 1, Alabama, looks like this:

S_cows_1

This is an awful lot of trades coming in. We would only expect an import into Alabama every 10 days on average.

Sure enough, if we alter the movement matrix, setting column 1 to all 0s (adding the previous values to a different column so rows still sum to 1), we see the same behaviour, despite saying that no trades should EVER have Alabama as a destination region.

S_cows2
(The axes are the wrong way round, but that's just me being daft)

An additional problem. We start losing cows from the system somehow. 
Total_cows1

This happens shortly after t=136, the point where trades are tested. This does not occur if we set time_test to somewhere far in the future, like 1e6.

total_cows_2

I now start diagnosing....

### CODE REVIEW

**Comment by thomrawson** [comment]
This was not a problem (I think) before  merging #7 . I have a screengrab of the trajectories for a bunch of states from before #7 was merged, and Alabama looks more normal here:
pr6_output1

**Comment by thomrawson** [comment]
The movement matrix bug was caused by [this addition](https://github.com/mrc-ide/cowflu/pull/7/commits/bbe14f672eab629ca625095992ef1c436305c966#diff-20d022f9ddeab0b3c10ef0b046094abb1bd07af27c883ffc262492ab8fdd0d62R206) in #7 .

It copied the above mechanism for defining i_region_dst, but forgets to add region_start, meaning a trade into herd 5 of any region, will always be sent to herd 5 of region 1.

[commit d2292d7](https://github.com/mrc-ide/cowflu/pull/11/commits/d2292d71e5d910a1ff55cf1f04ad32e52765a02f) fixes the bug. [commit 74ce516](https://github.com/mrc-ide/cowflu/pull/11/commits/74ce5166dbcbfc9eb7a540761e56ac5d9db7ef57) introduces a test to track it going forward.

**Comment by thomrawson** [comment]
[This commit](https://github.com/mrc-ide/cowflu/pull/11/commits/ad70a3e40a2f77751dca0b8a402366be925b1761) fixes an error in the number of hypergeometric draws made. This is how the testing of trades across borders is simulated.

Originally:
```
mcstate::random::hypergeometric(rng_state, internal.export_I[i_src], export_N - internal.export_I[i_src], std::min(shared.n_test, internal.export_I[i_src]))
```
`std::min(shared.n_test, internal.export_I[i_src])` is the number of draws, with n_test being the maximum number of cows that are tested in a cross-region herd export. But, we change it to `std::min(shared.n_test, export_N)`. If there are less than n_test cows being traded, then they ALL get tested, and any infections will be noticed (but not exposed, E). 

Consider these two examples which would happen with the previous code:

1) A herd is exporting 100 cows: 90 S, 9 E, and 1 I,
 
Previously, this hypergeometric will simulate sampling only 1 cow. When actually, we want it sample 30 (n_test). The idea being that if the one I is missed in the 30 (n_test) sampled, but some E cows go through, it seeds an infection in a new state.
 
2) A herd is exporting 20 cows - 10 S, 9 E, and 1 I.
 
Currently, it will draw 1 cow, risking infected cattle being traded. Exports with less than n_test cows get the entire export tested, and so any infected cattle in this number means the trade will not happen.

This bug is hard to write a test for, in theory it means that the number of trades occuring is greater than would be expected.

File: DESCRIPTION [hunk 1]
```diff
@@ -23,6 +23,7 @@ LinkingTo:
 Suggests:
     knitr,
     rmarkdown,
+    pkgload,
```
Comment (richfitz) [inline DESCRIPTION:26]:
```suggestion
```

I don't think we need this here?

File: inst/dust/cows.cpp [hunk 2]
```diff
@@ -205,15 +201,20 @@ class cows {
...
         if (allow_movement) {
           internal.import_S[i_dst] += internal.export_S[i_src];
           internal.import_E[i_dst] += internal.export_E[i_src];
           internal.import_I[i_dst] += internal.export_I[i_src];
           internal.import_R[i_dst] += internal.export_R[i_src];
+        } else {
+          internal.export_S[i_src] = 0;
+          internal.export_E[i_src] = 0;
+          internal.export_I[i_src] = 0;
+          internal.export_R[i_src] = 0;
```
Comment (richfitz) [inline inst/dust/cows.cpp:217]: these lines are fairly harmless, but I think they aren't strictly needed as we do the `std::fill()` above?

**Review by richfitz — APPROVED** [review]
nice sleuthing Thom!

File: inst/dust/cows.cpp (hunk 2 above)
Comment (thomrawson) [reply inst/dust/cows.cpp:217]: No I think they're actually crucial. It says that "if movement **isn't** allowed, then cancel the export". 
If these lines were removed, and movement was **not** allowed, then the import would not happen, but these exports would still get subtracted away.

File: DESCRIPTION (hunk 1 above)
Comment (thomrawson) [reply DESCRIPTION]: Yes, I think that was only there for when I first added the demo scripts. Should be fine to remove.

File: inst/dust/cows.cpp (hunk 2 above)
Comment (thomrawson) [reply inst/dust/cows.cpp:217]: It's the addition of these lines that stops cows disappearing from the system at lines 223-227

Comment (richfitz) [reply inst/dust/cows.cpp:217]: oh of course! Great work digging through this

**Pull request merged**